source/windows/secure_channel_tls_handler.c

```diff
@@ -1135,16 +1190,111 @@ static int s_do_application_data_decrypt(struct aws_channel_handler *handler) {
+                        (void *)handler,
+                        sc_handler->read_extra);
+                }
+            }
+
+            SecBuffer input_buffers2[] = {
+                [0] =
+                    {
+                        .pvBuffer = sc_handler->buffered_read_in_data_buf.buffer,
+                        .cbBuffer = (unsigned long)sc_handler->buffered_read_in_data_buf.len,
+                        .BufferType = SECBUFFER_TOKEN,
+                    },
+                [1] =
+                    {
+                        .pvBuffer = NULL,
+                        .cbBuffer = 0,
+                        .BufferType = SECBUFFER_EMPTY,
+                    },
+            };
+
+            SecBufferDesc input_bufs_desc = {
+                .ulVersion = SECBUFFER_VERSION,
+                .cBuffers = 2,
+                .pBuffers = input_buffers2,
+            };
+
+            SecBuffer output_buffers[3];
+            AWS_ZERO_ARRAY(output_buffers);
+            output_buffers[0].BufferType = SECBUFFER_TOKEN;
+            output_buffers[1].BufferType = SECBUFFER_ALERT;
+            output_buffers[2].BufferType = SECBUFFER_EMPTY;
+
+            SecBufferDesc output_buffers_desc = {
+                .ulVersion = SECBUFFER_VERSION,
+                .cBuffers = 3,
+                .pBuffers = output_buffers,
+            };
+            char server_name_cstr[256];
+            AWS_ZERO_ARRAY(server_name_cstr);
+            AWS_FATAL_ASSERT(sc_handler->server_name.len < sizeof(server_name_cstr));
+            memcpy(server_name_cstr, sc_handler->server_name.buffer, sc_handler->server_name.len);
+
+            status = InitializeSecurityContextA(
+                &sc_handler->creds,
+                &sc_handler->sec_handle,
+                (SEC_CHAR *)server_name_cstr,
+                sc_handler->ctx_req,
+                0,
+                0,
+                &input_bufs_desc,
+                0,
+                NULL,
+                &output_buffers_desc,
+                &sc_handler->ctx_ret_flags,
+                NULL);
+            if (status == SEC_E_OK) {
+                error = AWS_OP_SUCCESS;
+                if (input_buffers2[1].BufferType == SECBUFFER_EXTRA && input_buffers2[1].cbBuffer > 0 &&
+                    sc_handler->buffered_read_in_data_buf.len > input_buffers2[1].cbBuffer) {
+                    memmove(
+                        sc_handler->buffered_read_in_data_buf.buffer,
+                        (sc_handler->buffered_read_in_data_buf.buffer + sc_handler->buffered_read_in_data_buf.len) -
+                            input_buffers2[1].cbBuffer,
+                        input_buffers2[1].cbBuffer);
+                    sc_handler->buffered_read_in_data_buf.len = input_buffers2[1].cbBuffer;
+                    sc_handler->read_extra = input_buffers2[1].cbBuffer;
+                    continue;
+                }
+                break;

```

Comment:
I can't figure out why we don't want to try to decrypt extra data when `sc_handler->buffered_read_in_data_buf.len == input_buffers2[1].cbBuffer`. I think, a comment explaining this would be useful.